composer/callbacks/checkpoint_saver.py

```diff
@@ -39,6 +47,58 @@
 __all__ = ['CheckpointSaver']
 
 
+def _upload_symlink_file(
+    remote_backend_name: str,
+    backend_kwargs: dict[str, Any],
+    remote_symlink_file_name: str,
+    local_symlink_file_name: str,
+    num_attempts: int,
+    remote_checkpoint_file_names: list[str],
+    main_process_pid: int,
+    is_remote_upload_failed: multiprocessing.Event, # pyright: ignore[reportGeneralTypeIssues]
+    max_wait_time_in_seconds: int = 3600,
+    wait_before_next_try_in_seconds: float = 30,
+):
+    """Wait the checkpoint file uploading to finish and start symlink file uploading."""
+    start_time = time.time()
+    object_store = build_remote_backend(remote_backend_name, backend_kwargs)
+
+    for remote_file_name in remote_checkpoint_file_names:
+        while True:
+            if is_remote_upload_failed.is_set():
+                log.debug(f'Stop symlink uploading since the checkpoint files uploading failed')
+                return
+            # Return if parent process exits
+            try:
+                os.kill(main_process_pid, 0)
+            except OSError:
+                return
+            try:
+                object_store.get_object_size(remote_file_name)
+                break
+            except Exception as e:
+                if not isinstance(e, FileNotFoundError):
+                    log.debug(f'Got exception {type(e)}: {str(e)} when accessing remote file {remote_file_name}')
```

Comment:
If credentials aren't set properly, will this make it fail slowly?

Reply:
I see the validate_credentials function later, does that solve this?

Reply:
Yes!